packages/backend/src/tokens/request.ts

```diff
@@ -395,3 +485,93 @@ export const debugRequestState = (params: RequestState) => {
   const { isSignedIn, proxyUrl, reason, message, publishableKey, isSatellite, domain } = params;
   return { isSignedIn, proxyUrl, reason, message, publishableKey, isSatellite, domain };
 };
+
+// getActivationEntity determines if the given URL and settings indicate a desire to activate a specific organization or personal workspace.
+function getActivationEntity(url: URL, options: OrganizationSyncOptions): ActivatibleEntity | null {
+  // Check for personal workspace activation
+  if (options.personalWorkspacePatterns) {
+    let matcher: MatchFunction<object>;
+    try {
+      matcher = match(options.personalWorkspacePatterns);
+    } catch (e) {
+      console.error(
+        `Invalid personal workspace pattern "${options.personalWorkspacePatterns}": "${e}". See https://www.npmjs.com/package/path-to-regexp`,
+      );
+      return null;
+    }
+
+    let personalResult: boolean | MatchResult<object>;
+    try {
+      personalResult = matcher(url.pathname);
+    } catch (e) {
+      // Intentionally not logging the path to avoid potentially leaking anything sensitive
+      console.error(`Failed to apply personal workspace pattern "${options.personalWorkspacePatterns}" to a path`, e);
+      return null;
+    }
+
+    if (personalResult) {
+      return { type: 'personalWorkspace' };
+    }
+  }
+
+  // Check for organization activation
+  if (options.organizationPatterns) {
+    let matcher: MatchFunction<object>;
+    try {
+      matcher = match(options.organizationPatterns);
+    } catch (e) {
+      console.error(
+        `Invalid organization pattern "${options.organizationPatterns}": "${e}". See https://www.npmjs.com/package/path-to-regexp`,
+      );
+      return null;
+    }
+
+    let orgResult: boolean | MatchResult<object>;
+    try {
+      orgResult = matcher(url.pathname);
+    } catch (e) {
+      // Intentionally not logging the path to avoid potentially leaking anything sensitive
+      console.error(`Failed to apply organization pattern "${options.organizationPatterns}" to a path`, e);
+      return null;
+    }
+
+    if (orgResult) {
+      const { slug = '', id = '' } = orgResult.params as { slug?: string; id?: string };
+      if (id) {
+        return { type: 'organization', organizationId: id };
+      }
+      if (slug) {
+        return { type: 'organization', organizationSlug: slug };
+      }
+      console.warn(
+        'Detected an organization pattern match, but no organization ID or slug was found in the URL. Does the pattern include `:id` or `:slug`?',
+      );
+    }
+  }
+  return null;
+}
+
+// ActivatibleEntity is an entity that can be activated by the handshake API.
+type ActivatibleEntity = {
```

Comment:
How about `OrganizationSyncTarget` or `OrganizationSyncEntity` - to be consistent with "organization sync" here.

Also, let's use `type` as a discriminator union here to make it really explicit that it's an impossible state to have `personalWorkspace` with `organizationId` or `organizationSlug`:

```ts
/** OrganizationSyncTarget is an entity that can be activated by the handshake API. **/
type OrganizationSyncTarget =
  | {
      type: 'personalWorkspace';
    }
  | {
      type: 'organization';
      organizationId: string;
    }
  | {
      type: 'organization';
      organizationSlug: string;
    };
```

It's also not possible to have both slug and ID at the same time, so let's keep those separate.